### PR TITLE
Implemented support for configuring up to 10 data filters for query and bucket level monitors using the visual editor.

### PR DESCRIPTION
**Pull request by AWSHurneyt**
### Description
Implemented support for configuring up to 10 data filters for query and bucket level monitors using the visual editor.
 
### Issues Resolved
https://github.com/opensearch-project/alerting-dashboards-plugin/issues/432

### Screenshots
Screenshot 2023-03-29 at 2 39 15 PM
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting-dashboards-plugin/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by AWSHurneyt** [comment]
Request to update documentation to reflect this change.
https://github.com/opensearch-project/documentation-website/issues/3546

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/utils/whereHelpers.js [hunk 1]
```diff
@@ -47,12 +47,59 @@ export const displayText = (whereValues) => {
...
+  if (_.isEmpty(value.toString())) return 'Required.';
+  if (filter.fieldRangeEnd < value) return 'Start should be less than end range.';
+  if (value < filter.fieldRangeStart) return 'End should be greater than start range.';
+  return undefined;
+};
+
+export const validateWhereFilter = (filter = FORMIK_INITIAL_WHERE_EXPRESSION_VALUES) => {
+  const fieldName = _.get(
+    filter,
+    'fieldName[0]',
```
Comment (lezzago) [inline public/pages/CreateMonitor/components/MonitorExpressions/expressions/utils/whereHelpers.js:60]: why only check the first value?

File: public/pages/CreateTrigger/containers/CreateTrigger/utils/formikToTrigger.js [hunk 2]
```diff
@@ -334,15 +335,17 @@ export function getResultsPath(isCount) {
...
-  if (fieldName.length > 0) {
-    composite_agg_filter[where.fieldName[0].label] = {
-      [where.operator]: where.fieldValue,
-    };
-    return composite_agg_filter;
-  }
+  filters.forEach((filter) => {
+    const fieldName = _.get(filter, 'fieldName', FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.fieldName);
+    if (fieldName.length > 0) {
+      composite_agg_filter[fieldName[0].label] = {
```
Comment (lezzago) [inline public/pages/CreateTrigger/containers/CreateTrigger/utils/formikToTrigger.js:343]: Why do we only get the first value?

**Review by lezzago — CHANGES_REQUESTED** [review]

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/utils/whereHelpers.js (hunk 1 above)
Comment (AWSHurneyt) [reply public/pages/CreateMonitor/components/MonitorExpressions/expressions/utils/whereHelpers.js:60]: The combo box object returns an array of selected objects, even when the `singleSelection` setting for the box is `true`.

File: public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js [hunk 3]
```diff
@@ -6,11 +6,20 @@
...
 import { MONITOR_TYPE } from '../../../../../utils/constants';
 import { QUERY_OPERATORS } from '../../../../Dashboard/components/FindingsDashboard/findingsUtils';
+import { ALLOWED_TYPES } from '../../../components/MonitorExpressions/expressions/WherePopover';
 
 export const BUCKET_COUNT = 5;
 
 export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);
 
+export const FORMIK_INITIAL_WHERE_EXPRESSION_VALUES = {
+  fieldName: [{ label: '', type: ALLOWED_TYPES[0] }], // This is an array because the EuiCombobox returns an array of {label: string, type: string} objects.
```
[the remark this answers is not in the document]
Comment (AWSHurneyt) [reply public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js:16]: The `typeof` the entries in the [ALLOWED_TYPES](https://github.com/opensearch-project/alerting-dashboards-plugin/blob/6dc848e32f82b7398356fbc43a250872f11996aa/public/pages/CreateMonitor/components/MonitorExpressions/expressions/WhereExpression.js#L61) constant is `string` consisting of the word `number`. Those `{label: string, type: string}` objects form the `options` used by the combo boxes.

[the remark this answers is not in the document]
Comment (lezzago) [reply public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js:16]: got it

**Review by lezzago — DISMISSED** [review]

File: public/pages/CreateTrigger/containers/CreateTrigger/utils/formikToTrigger.js [hunk 4]
```diff
@@ -334,15 +335,17 @@ export function getResultsPath(isCount) {
...
+        [filter.operator]: filter.fieldValue,
+      };
+    }
+  });
+  if (!_.isEmpty(composite_agg_filter)) return composite_agg_filter;
```
Comment (bowenlan-amzn) [inline public/pages/CreateTrigger/containers/CreateTrigger/utils/formikToTrigger.js:348]: Maybe a dumb question, what would be return if it's empty here

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/WhereExpression.js [hunk 5]
```diff
@@ -58,262 +27,117 @@ const propTypes = {
...
+            data-test-subj={`${fieldPath}addFilterButton`}
+            onClick={() => {
+              openExpression(Expressions.WHERE);
+              filtersArrayHelpers.push(
+                useTriggerFieldOperators
+                  ? {
+                      ...FORMIK_INITIAL_WHERE_EXPRESSION_VALUES,
+                      operator: TRIGGER_OPERATORS_MAP.INCLUDE,
+                    }
+                  : FORMIK_INITIAL_WHERE_EXPRESSION_VALUES
```
Comment (amsiglan) [inline public/pages/CreateMonitor/components/MonitorExpressions/expressions/WhereExpression.js:96]: Should this still be a new object?
`{ ...FORMIK_INITIAL_WHERE_EXPRESSION_VALUES }`

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      isOpen: false,
+    };
+  }
+
+  componentDidMount() {
+    this.setState({ isOpen: this.props.openedState });
+  }
```
Comment (amsiglan) [inline public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js:48]: We don't need to set this state after mounting, it is fine to do it in the constructor

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  FormikComboBox,
+  FormikSelect,
+  FormikFieldNumber,
+  FormikFieldText,
+} from '../../../../../components/FormControls';
+import { getFilteredIndexFields, getIndexFields } from './utils/dataTypes';
+import { TRIGGER_COMPARISON_OPERATORS } from '../../../../CreateTrigger/containers/DefineBucketLevelTrigger/DefineBucketLevelTrigger';
+import { DATA_TYPES } from '../../../../../utils/constants';
+
+export const ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];
```
Comment (amsiglan) [inline public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js:36]: If this is also used in other places, I would suggest moving it to a constants file

File: public/pages/CreateTrigger/containers/CreateTrigger/utils/formikToTrigger.js (hunk 4 above)
Comment (AWSHurneyt) [reply public/pages/CreateTrigger/containers/CreateTrigger/utils/formikToTrigger.js:348]: If `composite_agg_filter` is empty, the function will return `undefined`; which means the trigger condition will be generated without a `composite_agg_filter` attribute.

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/utils/whereHelpers.js [hunk 8]
```diff
@@ -47,12 +47,59 @@ export const displayText = (whereValues) => {
...
+      filterIsValid = filterIsValid && !_.isEmpty(filter.fieldValue?.toString());
+      break;
+    case OPERATORS_MAP.IN_RANGE:
+    case OPERATORS_MAP.NOT_IN_RANGE:
+      // These operators store the query values in the 'fieldRangeStart' and 'fieldRangeEnd'
+      // attributes of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
+      // Both of those values need to be validated.
+      filterIsValid =
+        filterIsValid &&
+        validateRange(filter.fieldRangeStart, filter) &&
```
[the remark this answers is not in the document]
Comment (AWSHurneyt) [reply public/pages/CreateMonitor/components/MonitorExpressions/expressions/utils/whereHelpers.js:88]: Thank you for catching that! Revised.

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/WhereExpression.js (hunk 5 above)
Comment (AWSHurneyt) [reply public/pages/CreateMonitor/components/MonitorExpressions/expressions/WhereExpression.js:96]: Good point. Revised.

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js (hunk 6 above)
Comment (AWSHurneyt) [reply public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js:48]: Good point. Revised.

File: public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js (hunk 7 above)
Comment (AWSHurneyt) [reply public/pages/CreateMonitor/components/MonitorExpressions/expressions/WherePopover.js:36]: Good point. Revised.

**Review by lezzago — DISMISSED** [review]

**Review by amsiglan — APPROVED** [review]

**Review by bowenlan-amzn — APPROVED** [review]

**Review by lezzago — APPROVED** [review]

**Pull request merged**